### PR TITLE
[Core] Optimize allocation and MPI management in #11719

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

This PR tries to optimize the allocation and improve https://github.com/KratosMultiphysics/Kratos/issues/11719. We maybe require to add asyncronous interface to the data communicator in order to properly fix the missing changes.

**🆕 Changelog**

- [Refactor SpatialSearchResultContainer](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/4c46d470334a2ef8179007d3d41974cd7965968b)
- [Refactor SpatialSearchResultContainerVector (WIP)](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/3c79e77cefb3dc7a978a6bb695604b3c89856971)
- [Update SpatialSearchResultContainer tests](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/edb1f82b8c7999fc00f22330fb7e20f26947070d)
- [Update SpatialSearchResultContainerVector tests](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/7def12ae8ed22c3da2008196021c13d2e5736d34)
- [Refactor SearchWrapper](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/ad4f520ca1c6cdec4a652d4a82a8e7b63f0023ed)
- [Expose new and missing methods](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/554bfb555bee6540a2513c71790ca090c27763f4)
- [Update MPI test](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/449deba3ab9e4433a5a63c3c77a6cc0bd2216fe4)
- [Refactor tests](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/48741b2e8d9acc475c30f95f47edcd2950b30567)
- [Remove maps, pure vectors, suggested interface by](https://github.com/KratosMultiphysics/Kratos/pull/11857/commits/2bab9dc610e1e9a3d005098f95f52c5727bb6bfd) @pooyan-dadvand
- Changes from https://github.com/KratosMultiphysics/Kratos/pull/11875

### CODE REVIEW

**Comment by loumalouomega** [comment]
@pooyan-dadvand this is the modification I mentioned earlier, I don't add it directly to the PR branch of https://github.com/KratosMultiphysics/Kratos/pull/11719. This PR should be reviewed and if approved, should be done before than https://github.com/KratosMultiphysics/Kratos/pull/11719

**Comment by pooyan-dadvand** [comment]
I think unordered_set can easily outperform any ordered list with binary search. (I remember it from some comparisons I read years ago but I don't have the link now)

So I would simply use unordered_set instead.

**Comment by loumalouomega** [comment]
Closing until some refactors are done

File: kratos/spatial_containers/spatial_search_result_container_vector.cpp [hunk 1]
```diff
@@ -118,12 +90,12 @@ void SpatialSearchResultContainerVector<TObjectType>::InitializeResults(
...
-
-    // Check if index is initialized
-    const auto it_find = mMapIdsStored.find(Index);
-    return (it_find != mMapIdsStored.end());
+    // Check size
+    if (Index >= mPointResults.size()) {
+        return false;
+    } else {
+        return true;
+    }
```
Comment (roigcarlo) [inline kratos/spatial_containers/spatial_search_result_container_vector.cpp]:
```suggestion
    return Index < mPointResults.size();
```

**Review by roigcarlo — APPROVED** [review]
:+1:

File: kratos/spatial_containers/spatial_search_result_container_vector.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/spatial_containers/spatial_search_result_container_vector.cpp]: Yes

**Comment by loumalouomega** [comment]
I will merge this and the following changes will be done in an independent branch and merged to the main one as this pR intended

**Pull request merged**